packages/openspp/src/Adaptor.js

```diff
@@ -0,0 +1,705 @@
+import {
+  execute as commonExecute,
+  composeNextState,
+  expandReferences,
+} from "@openfn/language-common";
+
+import pkg from "odoo";
+const { Odoo } = pkg;
+
+var sppConnector = null;
+
+
+export class Log {
+  static info(message) {
+    return console.info(`ℹ `, message);
+  }
+
+  static success(message) {
+    return console.info(`✓ Success at ${new Date()}:\n∟`, message);
+  }
+
+  static warn(message) {
+    return console.warn(`⚠ Warning at ${new Date()}:\n∟`, message);
+  }
+
+  static error(message) {
+    return console.error(`✗ Error at ${new Date()}:\n∟`, message);
+  }
+}
+
+
+/**
+ * Execute a sequence of operations.
+ * Wraps `language-common/execute` to make working with this API easier.
+ * @example
+ * execute(
+ *   create("foo"),
+ *   delete("bar")
+ * )(state)
+ * @private
+ * @param {Operations} operations - Operations to be performed.
+ * @returns {Operation}
+ */
+export function execute(...operations) {
+  sppConnector = null;
+
+  const initialState = {
+    references: [],
+    data: null,
+  };
+
+  return state => {
+    return commonExecute(...operations)({ ...initialState, ...state });
+  };
+}
+
+/**
+ * Logs in to OpenSpp, gets a session token.
+ * @example
+ *  login(state)
+ * @private
+ * @param {State} state - Runtime state.
+ * @returns {State}
+ */
+function login(state) {
+  const {baseUrl, username, password, database} = state.configuration;
+  sppConnector = new Odoo({
+    host: baseUrl,
+    database: database,
+    username: username,
+    password: password
+  });
+  sppConnector.connect((err) => {
+    if (err) {
+      Log.error(err);
+      throw new Error("Can't login to OpenSPP, please check your credentials or network!");
+    }
+  });
+  return state;
+}
+
+/**
+ * Create a brand new program membership for registrant.
+ * @example
+ *  _creatingEnrolledRegistrantToProgram("IND_Q4VGGZPF", "PROG_2023_00000001")
+ * @private
+ * @param {string} registrant_id - registrant_id of group / individual wanted to unenroll 
+ * @param {string} program_id - program_id of program 
+ */
+function _creatingEnrolledRegistrantToProgram(registrant_id, program_id) {
+  if (sppConnector === null) {
```

Comment:
here, the standard protocol is to perform your login when you define your own custom execute function up on line 53. See [salesforce](https://github.com/OpenFn/adaptors/blob/main/packages/salesforce/src/Adaptor.js#L712-L733) (or any of the database-connection adaptors) for good examples here.

FYI @josephjclark , this is a really interesting approach. how does this dovetail with your thoughts around shared scope in multi-tenant nodejs worker environments?

Reply:
Hi, yes, the connector should be initialised at the start of execute. Line 45 on my diff right now. The connector is actually initialised to null right now but unless there's a very good reason, I'd like to see the login happen there.

I like the pattern of declaring the connector in a top-level variable (regarding multi-tenant environments @taylordowns2000 my thinking has settled on: every workflow must run in a cleanly initialised adaptor scope, and it's the engine's responsibility to guarantee that (as our new engine is doing right now). Security first, stability second, performance third.

Also @nhatnm0612 a couple of small style points:
1) We prefer not to use underscored function names. This is marked as `@private`, that's good enough for me.
2) camelCased variable names are preferred to snake_cake. Data passed straight through to the backend or URL is exempt, but parameter names should really be camelCase to ensure a nice consistent presentation across adaptors.
3) Feel free to use an async/await pattern to simplify your code and remove the promise chain! I'd be happy to show you how to do this if you're not familiar with the style.

Reply:
@josephjclark @taylordowns2000  hi, I changed the style from callback to async await, can you re-review it again?